uvicorn/config.py

```diff
@@ -224,13 +227,57 @@ def __init__(
         self.loaded = False
         self.configure_logging()
 
-        if reload_dirs is None:
-            self.reload_dirs = [os.getcwd()]
-        else:
-            if isinstance(reload_dirs, str):
-                self.reload_dirs = [reload_dirs]
-            else:
-                self.reload_dirs = reload_dirs
+        self.reload_includes: List[str] = []
+        self.reload_excludes: List[str] = []
+        self.reload_dirs_excludes: List[Path] = []
+        reload_dirs_list = list(set(reload_dirs)) if reload_dirs else []
+        reload_includes = list(set(reload_includes if reload_includes else []))
+        reload_excludes = list(set(reload_excludes if reload_excludes else []))
+
+        for pattern in reload_includes:
+            try:
+                if Path(pattern).is_dir():
+                    reload_dirs_list.append(pattern)
+                    reload_includes.remove(pattern)
+                else:
+                    self.reload_includes.append(pattern)
+            except OSError:
+                self.reload_includes.append(pattern)
+
+        for pattern in reload_excludes:
+            try:
+                if Path(pattern).is_dir():
+                    self.reload_dirs_excludes.append(Path(pattern))
+                else:
+                    self.reload_excludes.append(pattern)
+            except OSError:
+                self.reload_excludes.append(pattern)
+
+        reload_dirs_list = list(set(reload_dirs_list))
+        reload_paths = list(map(Path, reload_dirs_list))
+        reload_paths = list(map(lambda x: x.resolve(), reload_paths))
+        reload_paths = list(
+            set([reload_path for reload_path in reload_paths if reload_path.is_dir()])
+        )
+
+        if not reload_paths:
+            if reload_dirs:
+                logger.warning(
+                    "Provided reload directories %s did not contain valid "
+                    + "directories, watching current working directory.",
+                    reload_dirs,
+                )
+            reload_paths = [Path(os.getcwd())]
+
+        children = []
+        for j in range(len(reload_paths)):
+            for k in range(j + 1, len(reload_paths)):
+                if reload_paths[j] in reload_paths[k].parents:
+                    children.append(reload_paths[k])
+                elif reload_paths[k] in reload_paths[j].parents:
+                    children.append(reload_paths[j])
```

Comment:
seems like 278 is not covered,
if I do a 
```
        children = []
        for j in range(len(reload_paths)):
            print(0)
            for k in range(j + 1, len(reload_paths)):
                if reload_paths[j] in reload_paths[k].parents:
                    print(1)
                    children.append(reload_paths[k])
                elif reload_paths[k] in reload_paths[j].parents:
                    print(2)
                    children.append(reload_paths[j])
```
and run the test suite it seems no 1 or 2 are printed, I'm not sure I get what this children is doing, would you show a test case that demonstrate its use ?

Reply:
The use case is to find directories which are children of other watched directories.
- app
--- sub
and both app and sub are added to reloaddirs.
It would be a waste to create watchers for both app and sub, therefore I remove the children.
I have added a [new test case](https://github.com/Roang-zero1/uvicorn/commit/9a10bf80f7077ea9e78ca7cefe0d0217746a9ae6#diff-da24b355349b53e4c8047034767d008e4c2ce9ea6c8933a3acd50d0828f00528R132-R151), unfortunately I am [using a set](https://github.com/Roang-zero1/uvicorn/blob/9a10bf80f7077ea9e78ca7cefe0d0217746a9ae6/uvicorn/config.py#L255-L260) beforehand so I can't guarantee that both branches are met every time...
@euri10 Do you have input on how I could solve this in the test?

Reply:
no this looks fine this way